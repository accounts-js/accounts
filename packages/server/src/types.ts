import { UserObjectType, CreateUserType, SessionType } from '@accounts/common';
import { EmailTemplatesType, EmailTemplateType, SendMailType } from './email';
import { AccountsServer } from './accounts-server';

export interface AuthService {
  server: AccountsServer;
  serviceName: string;
  setStore(store: DBInterface): void;
  authenticate(params: any): Promise<UserObjectType | null>;
}

export type UserObjectSanitizerFunction = (
  userObject: UserObjectType,
  omitFunction: (userDoc: object, fields: string[]) => UserObjectType,
  pickFunction: (userDoc: object, fields: string[]) => UserObjectType
) => any;

export type ResumeSessionValidator = (user: UserObjectType, session: SessionType) => Promise<any>;

export type PrepareMailFunction = (
  to: string,
  token: string,
  user: UserObjectType,
  pathFragment: string,
  emailTemplate: EmailTemplateType,
  from: string
) => object;

export type EmailType = EmailTemplateType & { to: string };

export interface AccountsServerOptions {
  db: DBInterface;
  tokenSecret: string;
  tokenConfigs?: {
    accessToken?: {
      expiresIn?: string;
    };
    refreshToken?: {
      expiresIn?: string;
    };
  };
  emailTokensExpiry?: number;
  emailTemplates?: EmailTemplatesType;
  userObjectSanitizer?: UserObjectSanitizerFunction;
  impersonationAuthorize?: (
    user: UserObjectType,
    impersonateToUser: UserObjectType
  ) => Promise<any>;
  resumeSessionValidator?: ResumeSessionValidator;
  siteUrl?: string;
  prepareMail?: PrepareMailFunction;
  sendMail?: SendMailType;
  // https://github.com/eleith/emailjs#emailserverconnectoptions
  email?: object;
}

export interface ConnectionInformationsType {
  ip?: string;
  userAgent?: string;
}

export interface DBInterface {
  // Find user by identity fields
  findUserByEmail(email: string): Promise<UserObjectType | null>;
  findUserByUsername(username: string): Promise<UserObjectType | null>;
  findUserById(userId: string): Promise<UserObjectType | null>;

  // Create and update users
  createUser(user: CreateUserType): Promise<string>;
  setUsername(userId: string, newUsername: string): Promise<void>;
  setProfile(userId: string, profile: object): Promise<object>;

  // Auth services related operations
  findUserByServiceId(serviceName: string, serviceId: string): Promise<UserObjectType | null>;
  setService(userId: string, serviceName: string, data: object): Promise<void>;
  unsetService(userId: string, serviceName: string): Promise<void>;

  // Password related operation
  findPasswordHash(userId: string): Promise<string | null>;
  findUserByResetPasswordToken(token: string): Promise<UserObjectType | null>;
  setPassword(userId: string, newPassword: string): Promise<void>;
  addResetPasswordToken(
    userId: string,
    email: string,
    token: string,
    reason?: string
  ): Promise<void>;
  setResetPassword(
    userId: string,
    email: string,
    newPassword: string,
    token: string
  ): Promise<void>;

  // Email related operations
  findUserByEmailVerificationToken(token: string): Promise<UserObjectType | null>;
  addEmail(userId: string, newEmail: string, verified: boolean): Promise<void>;
  removeEmail(userId: string, email: string): Promise<void>;
  verifyEmail(userId: string, email: string): Promise<void>;
  addEmailVerificationToken(userId: string, email: string, token: string): Promise<void>;

  // Session related operations
  findSessionById(sessionId: string): Promise<SessionType | null>;
  findSessionByToken(token: string): Promise<SessionType | null>;
  createSession(
    userId: string,
    token: string,
    connection: ConnectionInformationsType,
    extraData?: object
  ): Promise<string>;
  updateSession(sessionId: string, connection: ConnectionInformationsType): Promise<void>;
  invalidateSession(sessionId: string): Promise<void>;
  invalidateAllSessions(userId: string): Promise<void>;
}
