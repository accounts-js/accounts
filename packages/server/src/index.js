/* eslint-disable import/no-named-as-default */
import Accounts, { AccountsServer } from './AccountsServer';
import * as encryption from './encryption';
import type { DBInterface } from './DBInterface';
import config from './config';

export default Accounts;

export {
  AccountsServer,
  encryption,
  config,
};

export type {
  DBInterface,
};
