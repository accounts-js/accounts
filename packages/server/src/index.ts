import { AccountsServer, AuthenticationServices } from './accounts-server';
export { AccountsServerOptions } from './types/accounts-server-options';
export { generateRandomToken } from './utils/tokens';
export { getFirstUserEmail } from './utils/get-first-user-email';
export { ServerHooks } from './utils/server-hooks';
export { AccountsJsError } from './utils/accounts-error';
export {
  AuthenticateWithServiceErrors,
  LoginWithServiceErrors,
  ImpersonateErrors,
  FindSessionByAccessTokenErrors,
  RefreshTokensErrors,
  LogoutErrors,
  ResumeSessionErrors,
} from './errors';

export default AccountsServer;
export { AccountsServer, AuthenticationServices };
