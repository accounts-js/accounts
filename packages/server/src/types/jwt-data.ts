export interface JwtData {
  token: string;
  isImpersonated: boolean;
  userId: string;
}
