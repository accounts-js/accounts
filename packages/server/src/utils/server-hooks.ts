export const ServerHooks = {
  ValidateLogin: 'ValidateLogin',
  LoginSuccess: 'LoginSuccess',
  LoginError: 'LoginError',
  AuthenticateSuccess: 'AuthenticateSuccess',
  AuthenticateError: 'AuthenticateError',
  LogoutSuccess: 'LogoutSuccess',
  LogoutError: 'LogoutError',
  CreateUserSuccess: 'CreateUserSuccess',
  CreateUserError: 'CreateUserError',
  ResumeSessionSuccess: 'ResumeSessionSuccess',
  ResumeSessionError: 'ResumeSessionError',
  RefreshTokensSuccess: 'RefreshTokensSuccess',
  RefreshTokensError: 'RefreshTokensError',
  ImpersonationSuccess: 'ImpersonationSuccess',
  ImpersonationError: 'ImpersonationError',
  ResetPasswordSuccess: 'ResetPasswordSuccess',
  ChangePasswordSuccess: 'ChangePasswordSuccess',
};
