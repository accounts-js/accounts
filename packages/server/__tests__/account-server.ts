import * as jwtDecode from 'jwt-decode';
import { AccountsServer } from '../src/accounts-server';
import { JwtData } from '../src/types/jwt-data';
import { ServerHooks } from '../src/utils/server-hooks';

describe('AccountsServer', () => {
  const db = {
    findUserByUsername: () => Promise.resolve(),
    findUserByEmail: () => Promise.resolve(),
    createUser: () => Promise.resolve(),
    createSession: () => Promise.resolve(),
  };

  describe('config', () => {
    it('throws on invalid db', async () => {
      try {
        const account = new AccountsServer({} as any, {});
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });
  });

  describe('getServices', () => {
    it('should return instance services', async () => {
      const services: any = {
        password: {
          setStore: () => null,
        },
      };
      const account = new AccountsServer({ db: {} } as any, services);
      expect(account.getServices()).toEqual(services);
    });
  });

  describe('loginWithService', () => {
    it('throws on invalid service', async () => {
      try {
        const accountServer = new AccountsServer({ db: {} } as any, {});
        await accountServer.loginWithService('facebook', {}, {});
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('throws when user not found', async () => {
      const authenticate = jest.fn(() => Promise.resolve());
      try {
        const accountServer = new AccountsServer({ db: {} } as any, {
          facebook: { authenticate, setStore: jest.fn() },
        });
        await accountServer.loginWithService('facebook', {}, {});
        throw new Error();
      } catch (err) {
        expect(err.message).toMatchSnapshot();
      }
    });

    it('should return tokens', async () => {
      const authenticate = jest.fn(() => Promise.resolve({ id: 'userId' }));
      const createSession = jest.fn(() => Promise.resolve('sessionId'));
      const accountServer = new AccountsServer(
        {
          db: { createSession } as any,
          tokenSecret: 'secret1',
        },
        {
          facebook: { authenticate, setStore: jest.fn() },
        }
      );
      const res = await accountServer.loginWithService('facebook', {}, {});
      expect(res.tokens).toBeTruthy();
    });
  });

  describe('loginWithUser', () => {
    it('creates a session when given a proper user object', async () => {
      const user = {
        id: '123',
        username: 'username',
        email: 'email@email.com',
        profile: {
          bio: 'bio',
        },
      };
      const accountsServer = new AccountsServer(
        {
          db: {
            createSession: () => Promise.resolve('sessionId'),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      const res = await accountsServer.loginWithUser(user, {});
      const { accessToken, refreshToken } = res.tokens;
      const decodedAccessToken: { data: JwtData } = jwtDecode(accessToken);
      expect(decodedAccessToken.data.token).toBeTruthy();
      expect(accessToken).toBeTruthy();
      expect(refreshToken).toBeTruthy();
    });
  });

  describe('logout', () => {
    it('invalidates session', async () => {
      const invalidateSession = jest.fn(() => Promise.resolve());
      const user = {
        userId: '123',
        username: 'username',
      };

      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            invalidateSession,
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      const { accessToken } = accountsServer.createTokens('456');
      await accountsServer.logout(accessToken);
      expect(invalidateSession).toBeCalledWith('456');
    });
  });

  describe('hooks', () => {
    const connectionInfo = {
      userAgent: 'user-agent-test',
      ip: 'ip-test',
    };
    it('ServerHooks.LoginSuccess', async () => {
      const user = { id: 'id-test' };
      const hookSpy = jest.fn(() => null);
      const services = {
        password: {
          setStore: jest.fn(),
          authenticate: jest.fn(() => Promise.resolve(user)),
        },
      };
      const accountsServer = new AccountsServer(
        {
          db: {
            createSession: () => '123',
          } as any,
          tokenSecret: 'secret1',
        },
        services
      );
      accountsServer.on(ServerHooks.LoginSuccess, hookSpy);
      await accountsServer.loginWithService('password', { key: 'value' }, connectionInfo);
      expect(hookSpy).toHaveBeenCalledWith({
        service: 'password',
        connection: connectionInfo,
        user,
        params: { key: 'value' },
      });
    });

    it('ServerHooks.LoginError', async () => {
      const user = { id: 'id-test' };
      const services = {
        password: {
          setStore: jest.fn(),
          authenticate: jest.fn(() => Promise.resolve(user)),
        },
      };
      const hookSpy = jest.fn(() => null);
      const SessionError = new Error('Could not create session');
      const accountsServer = new AccountsServer(
        {
          db: {
            createSession: () => {
              throw SessionError;
            },
          } as any,
          tokenSecret: 'secret1',
        },
        services
      );
      accountsServer.on(ServerHooks.LoginError, hookSpy);

      try {
        await accountsServer.loginWithService('password', { key: 'value' }, connectionInfo);
      } catch (e) {
        // nothing to do
      }
      expect(hookSpy).toHaveBeenCalledWith({
        service: 'password',
        connection: connectionInfo,
        user,
        params: { key: 'value' },
        error: SessionError,
      });
    });

    it('ServerHooks.LogoutSuccess', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };

      const hookSpy = jest.fn(() => null);
      const invalidateSession = jest.fn(() => Promise.resolve());
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
            invalidateSession,
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      accountsServer.on(ServerHooks.LogoutSuccess, hookSpy);

      const { accessToken } = accountsServer.createTokens('456');
      await accountsServer.logout(accessToken);
      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.LogoutError', async () => {
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      accountsServer.on(ServerHooks.LogoutError, hookSpy);

      try {
        const { accessToken } = accountsServer.createTokens('456');
        await accountsServer.logout(accessToken);
      } catch (err) {
        // nothing to do
      }

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.ResumeSessionSuccess', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
            resumeSessionValidator: () => Promise.resolve(user),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      accountsServer.on(ServerHooks.ResumeSessionSuccess, hookSpy);

      const { accessToken } = accountsServer.createTokens('456');
      await accountsServer.resumeSession(accessToken);

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.ResumeSessionError with invalid session', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: false,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
          } as any,
          tokenSecret: 'secret1',
          resumeSessionValidator: () => Promise.resolve(user),
        },
        {}
      );
      accountsServer.on(ServerHooks.ResumeSessionError, hookSpy);

      const { accessToken } = accountsServer.createTokens('456');

      try {
        await accountsServer.resumeSession(accessToken);
      } catch (e) {
        // nothing to do
      }

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.ResumeSessionError with invalid errored session', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () => Promise.reject(''),
            findUserById: () => Promise.resolve(user),
          } as any,
          tokenSecret: 'secret1',
          resumeSessionValidator: () => Promise.resolve(user),
        },
        {}
      );
      accountsServer.on(ServerHooks.ResumeSessionError, hookSpy);

      const { accessToken } = accountsServer.createTokens('456');

      try {
        await accountsServer.resumeSession(accessToken);
      } catch (e) {
        // nothing to do
      }

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.RefreshTokensError', async () => {
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                valid: false,
              }),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      accountsServer.on(ServerHooks.RefreshTokensError, hookSpy);

      try {
        const { accessToken, refreshToken } = accountsServer.createTokens(null);
        await accountsServer.refreshTokens(accessToken, refreshToken, null, null);
      } catch (err) {
        // nothing to do
      }

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.RefreshTokensSuccess', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
            updateSession: () => Promise.resolve(),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      accountsServer.on(ServerHooks.RefreshTokensSuccess, hookSpy);

      const { accessToken, refreshToken } = accountsServer.createTokens('456');
      accountsServer.createTokens = () => ({
        accessToken: 'newAccessToken',
        refreshToken: 'newRefreshToken',
      });

      await accountsServer.refreshTokens(accessToken, refreshToken, 'ip', 'user agent');

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.ImpersonationError', async () => {
      const hookSpy = jest.fn(() => null);
      const accountsServer = new AccountsServer(
        {
          db: db as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      accountsServer.on(ServerHooks.ImpersonationError, hookSpy);

      try {
        await accountsServer.impersonate(null, null, null, null);
      } catch (err) {
        // nothing to do
      }

      expect(hookSpy).toBeCalled();
    });

    it('ServerHooks.ImpersonationSuccess', async () => {
      const user = { username: 'myUser', id: '123' };
      const impersonatedUser = { username: 'impUser', id: '456' };
      const findUserById = jest.fn();
      findUserById.mockReturnValueOnce(user).mockReturnValueOnce(impersonatedUser);
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById,
            createSession: () => Promise.resolve('001'),
          } as any,
          tokenSecret: 'secret1',
          impersonationAuthorize: async (userObject, impersonateToUser) => {
            return userObject.id === user.id && impersonateToUser === impersonatedUser;
          },
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');
      const hookSpy = jest.fn(() => null);
      accountsServer.on(ServerHooks.ImpersonationSuccess, hookSpy);

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: true,
          userId: '123',
        } as any);
      accountsServer.createTokens = (sessionId, isImpersonated) =>
        ({
          sessionId,
          isImpersonated,
        } as any);

      await accountsServer.impersonate(accessToken, { userId: 'userId' }, null, null);

      expect(hookSpy).toBeCalled();
    });
  });

  describe('refreshTokens', () => {
    it('updates session and returns new tokens and user', async () => {
      const updateSession = jest.fn(() => Promise.resolve());
      const user = {
        userId: '123',
        username: 'username',
      };
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
            updateSession,
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const { accessToken, refreshToken } = accountsServer.createTokens('456');
      accountsServer.createTokens = () => ({
        accessToken: 'newAccessToken',
        refreshToken: 'newRefreshToken',
      });
      const res = await accountsServer.refreshTokens(accessToken, refreshToken, 'ip', 'user agent');
      expect(updateSession.mock.calls[0]).toEqual(['456', { ip: 'ip', userAgent: 'user agent' }]);
      expect(res.user).toEqual({
        userId: '123',
        username: 'username',
      });
    });

    it('requires access and refresh tokens', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {} as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        await accountsServer.refreshTokens(null, null, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('An accessToken and refreshToken are required');
      }
    });
    it('throws error if tokens are not valid', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {} as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        await accountsServer.refreshTokens('bad access token', 'bad refresh token', null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Tokens are not valid');
      }
    });

    it('throws error if session not found', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        const { accessToken, refreshToken } = accountsServer.createTokens(null);
        await accountsServer.refreshTokens(accessToken, refreshToken, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Session not found');
      }
    });

    it('throws error if session not valid', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                valid: false,
              }),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        const { accessToken, refreshToken } = accountsServer.createTokens(null);
        await accountsServer.refreshTokens(accessToken, refreshToken, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Session is no longer valid');
      }
    });

    it('throws error if user not found', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        const { accessToken, refreshToken } = accountsServer.createTokens(null);
        await accountsServer.refreshTokens(accessToken, refreshToken, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('User not found');
      }
    });
  });

  describe('findSessionByAccessToken', () => {
    it('requires access token', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {} as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        await accountsServer.logout(null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('An accessToken is required');
      }
    });
    it('throws error if tokens are not valid', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {} as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        await accountsServer.logout('bad access token');
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Tokens are not valid');
      }
    });
    it('throws error if session not found', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        const { accessToken } = accountsServer.createTokens(null);
        await accountsServer.logout(accessToken);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Session not found');
      }
    });
    it('throws error if session not valid', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                valid: false,
              }),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        const { accessToken } = accountsServer.createTokens(null);
        await accountsServer.logout(accessToken);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Session is no longer valid');
      }
    });
  });

  describe('findUserById', () => {
    it('call this.db.findUserById', async () => {
      const findUserById = jest.fn(() => Promise.resolve('user'));
      const accountsServer = new AccountsServer(
        {
          db: { findUserById } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const user = await accountsServer.findUserById('id');
      expect(findUserById.mock.calls[0]).toEqual(['id']);
      expect(user).toEqual('user');
    });
  });

  describe('resumeSession', () => {
    it('throws error if user is not found', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      try {
        const { accessToken } = accountsServer.createTokens('456');
        await accountsServer.resumeSession(accessToken);
        throw new Error();
      } catch (err) {
        const { message } = err;
        expect(message).toEqual('User not found');
      }
    });

    it('should throw if session is not valid', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: false,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        const { accessToken } = accountsServer.createTokens('456');
        const ret = await accountsServer.resumeSession(accessToken);
      } catch (err) {
        const { message } = err;
        expect(message).toEqual('Invalid Session');
      }
    });

    it('return user', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const accountsServer = new AccountsServer(
        {
          db: {
            findSessionByToken: () =>
              Promise.resolve({
                id: '456',
                valid: true,
                userId: '123',
              }),
            findUserById: () => Promise.resolve(user),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      const { accessToken } = accountsServer.createTokens('456');
      const foundUser = await accountsServer.resumeSession(accessToken);
      expect(foundUser).toEqual(user);
    });
  });

  describe('setProfile', () => {
    it('throws error if user is not found', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        await accountsServer.setProfile(null, null);
        throw new Error();
      } catch (err) {
        const { message } = err;
        expect(message).toEqual('User not found');
      }
    });

    it('calls set profile on db interface', async () => {
      const user = {
        userId: '123',
        username: 'username',
      };
      const profile = {
        bio: 'bio',
      };
      const setProfile = jest.fn();
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById: () => Promise.resolve(user),
            setProfile,
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      await accountsServer.setProfile('123', profile);
      expect(setProfile.mock.calls.length).toEqual(1);
      expect(setProfile.mock.calls[0][0]).toEqual('123');
      expect(setProfile.mock.calls[0][1]).toEqual(profile);
    });

    it('merges profile and calls set profile on db interface', async () => {
      const user = {
        userId: '123',
        username: 'username',
        profile: {
          title: 'title',
        },
      };
      const profile = {
        bio: 'bio',
      };
      const mergedProfile = {
        title: 'title',
        bio: 'bio',
      };
      const setProfile = jest.fn(() => mergedProfile);
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById: () => Promise.resolve(user),
            setProfile,
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      const res = await accountsServer.updateProfile('123', profile);
      expect(setProfile.mock.calls.length).toEqual(1);
      expect(setProfile.mock.calls[0][0]).toEqual('123');
      expect(setProfile.mock.calls[0][1]).toEqual(mergedProfile);
      expect(res).toEqual(mergedProfile);
    });
  });

  describe('impersonate', () => {
    const user = { username: 'myUser', id: '123' };
    const impersonatedUser = { username: 'impUser', id: '456' };
    const someUser = { username: 'someUser', id: '789' };

    it('throws error if no access token is provided', async () => {
      const accountsServer = new AccountsServer(
        {
          db: db as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      try {
        await accountsServer.impersonate(null, null, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('An access token is required');
      }
    });

    it('throws error if access token is not valid', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {} as any,
          tokenSecret: 'secret1',
        },
        {}
      );

      try {
        const res = await accountsServer.impersonate('invalidToken', {}, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Access token is not valid');
      }
    });

    it('throws error if session is not valid', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {} as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: false,
          userId: '123',
        } as any);

      try {
        const res = await accountsServer.impersonate(accessToken, {}, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Session is not valid for user');
      }
    });

    it('throws error if user is not found', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: true,
          userId: '123',
        } as any);

      try {
        const res = await accountsServer.impersonate(accessToken, { userId: 'userId' }, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('User not found');
      }
    });

    it('throws error if impersonated user is not found', async () => {
      const findUserById = jest.fn();
      findUserById.mockReturnValueOnce(user);
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById,
            findUserByUsername: () => Promise.resolve(null),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: true,
          userId: '123',
        } as any);

      try {
        const res = await accountsServer.impersonate(accessToken, { userId: 'userId' }, null, null);
        throw new Error();
      } catch (err) {
        expect(err.message).toEqual('Impersonated user not found');
      }
    });

    it('returns not authorized if impersonationAuthorize function is not passed in config', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById: () => Promise.resolve(user),
            findUserByUsername: () => Promise.resolve(someUser),
          } as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: true,
          userId: '123',
        } as any);

      const res = await accountsServer.impersonate(accessToken, { userId: 'userId' }, null, null);
      expect(res.authorized).toEqual(false);
    });

    it('returns not authorized if impersonationAuthorize return false', async () => {
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById: () => Promise.resolve(user),
            findUserByUsername: () => Promise.resolve(someUser),
          } as any,
          tokenSecret: 'secret1',
          impersonationAuthorize: async (userObject, impersonateToUser) => {
            return userObject.id === user.id && impersonateToUser === impersonatedUser;
          },
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: true,
          userId: '123',
        } as any);

      const impersonationAuthorize = accountsServer.getOptions().impersonationAuthorize;
      expect(impersonationAuthorize).toBeDefined();

      const res = await accountsServer.impersonate(accessToken, { userId: 'userId' }, null, null);
      expect(res.authorized).toEqual(false);
    });

    it('returns correct response if authorized', async () => {
      const createSession = jest.fn(() => Promise.resolve('001'));
      const findUserById = jest.fn();
      findUserById.mockReturnValueOnce(user).mockReturnValueOnce(impersonatedUser);
      const accountsServer = new AccountsServer(
        {
          db: {
            findUserById,
            createSession,
          } as any,
          tokenSecret: 'secret1',
          impersonationAuthorize: async (userObject, impersonateToUser) => {
            return userObject.id === user.id && impersonateToUser === impersonatedUser;
          },
        },
        {}
      );
      const { accessToken } = accountsServer.createTokens('555');

      accountsServer.findSessionByAccessToken = () =>
        Promise.resolve({
          valid: true,
          userId: '123',
        } as any);
      accountsServer.createTokens = (sessionId, isImpersonated) =>
        ({
          sessionId,
          isImpersonated,
        } as any);

      const res = await accountsServer.impersonate(accessToken, { userId: 'userId' }, null, null);
      expect(res).toEqual({
        authorized: true,
        tokens: { sessionId: '001', isImpersonated: true },
        user: impersonatedUser,
      });
      expect(createSession.mock.calls[0][3]).toEqual({
        impersonatorUserId: user.id,
      });
    });
  });

  describe('user sanitizer', () => {
    const userObject = { username: 'test', services: [], id: '123' };

    it('internal sanitizer should clean services field from the user object', () => {
      const accountsServer = new AccountsServer(
        {
          db: db as any,
          tokenSecret: 'secret1',
        },
        {}
      );
      const modifiedUser = accountsServer.sanitizeUser(userObject);
      expect(modifiedUser.services).toBeUndefined();
    });

    it('should run external sanitizier when provided one', () => {
      const accountsServer = new AccountsServer(
        {
          db: db as any,
          tokenSecret: 'secret1',
          userObjectSanitizer: (user, omit) => omit(user, ['username']),
        },
        {}
      );
      const modifiedUser = accountsServer.sanitizeUser(userObject);
      expect(modifiedUser.username).toBeUndefined();
    });
  });
});
