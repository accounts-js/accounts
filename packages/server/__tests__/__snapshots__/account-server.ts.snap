// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsServer authenticateWithService throws on invalid service 1`] = `"No service with the name facebook was registered."`;

exports[`AccountsServer authenticateWithService throws when user is deactivated 1`] = `"Your account has been deactivated"`;

exports[`AccountsServer authenticateWithService throws when user not found 1`] = `"Service facebook was not able to authenticate user"`;

exports[`AccountsServer config throws on invalid db 1`] = `"A database driver is required"`;

exports[`AccountsServer loginWithService throws on invalid service 1`] = `"No service with the name facebook was registered."`;

exports[`AccountsServer loginWithService throws when user is deactivated 1`] = `"Your account has been deactivated"`;

exports[`AccountsServer loginWithService throws when user not found 1`] = `"Service facebook was not able to authenticate user"`;
